README.md
== Synacor Challenge ==
In this challenge, your job is to use this architecture spec to create a
virtual machine capable of running the included binary.  Along the way,
you will find codes; submit these to the challenge website to track
your progress.  Good luck!


== architecture ==
- three storage regions
  - memory with 15-bit address space storing 16-bit values
  - eight registers
  - an unbounded stack which holds individual 16-bit values
- all numbers are unsigned integers 0..32767 (15-bit)
- all math is modulo 32768; 32758 + 15 => 5

== binary format ==
- each number is stored as a 16-bit little-endian pair (low byte, high byte)
- numbers 0..32767 mean a literal value
- numbers 32768..32775 instead mean registers 0..7
- numbers 32776..65535 are invalid
- programs are loaded into memory starting at address 0
- address 0 is the first 16-bit value, address 1 is the second 16-bit value, etc

== execution ==
- After an operation is executed, the next instruction to read is immediately after the last argument of the current operation.  If a jump was performed, the next operation is instead the exact destination of the jump.
- Encountering a register as an operation argument should be taken as reading from the register or setting into the register as appropriate.

== hints ==
- Start with operations 0, 19, and 21.
- Here's a code for the challenge website: NRsyGxrDOwMn
- The program "9,32768,32769,4,19,32768" occupies six memory addresses and should:
  - Store into register 0 the sum of 4 and the value contained in register 1.
  - Output to the terminal the character with the ascii code contained in register 0.

== opcode listing ==
- halt: 0
  stop execution and terminate the program
- set: 1 a b
  set register < A > to the value of < B >
- push: 2 a
  push < A > onto the stack
- pop: 3 a
  remove the top element from the stack and write it into < A >; empty stack = error
- eq: 4 a b c
  set < A > to 1 if < B > is equal to <c>; set it to 0 otherwise
- gt: 5 a b c
  set < A > to 1 if < B > is greater than <c>; set it to 0 otherwise
- jmp: 6 a
  jump to < A >
- jt: 7 a b
  if < A > is nonzero, jump to < B >
- jf: 8 a b
  if < A > is zero, jump to < B >
- add: 9 a b c
  assign into < A > the sum of < B > and <c> (modulo 32768)
- mult: 10 a b c
  store into < A > the product of < B > and <c> (modulo 32768)
- mod: 11 a b c
  store into < A > the remainder of < B > divided by <c>
- and: 12 a b c
  stores into < A > the bitwise and of < B > and <c>
- or: 13 a b c
  stores into < A > the bitwise or of < B > and <c>
- not: 14 a b
  stores 15-bit bitwise inverse of < B > in < A >
- rmem: 15 a b
  read memory at address < B > and write it to < A >
- wmem: 16 a b
  write the value from < B > into memory at address < A >
- call: 17 a
  write the address of the next instruction to the stack and jump to < A >
- ret: 18
  remove the top element from the stack and jump to it; empty stack = halt
- out: 19 a
  write the character represented by ascii code < A > to the terminal
- in: 20 a
  read a character from the terminal and write its ascii code to < A >; it can be assumed that once input starts, it will continue until a newline is encountered; this means that you can safely read whole lines from the keyboard and trust that they will be fully read
- noop: 21
  no operation


changelog.txt
dev version - 16/12/2015
------------------------------
* code clean up
* function keys to the VM
	- trace implemented - now disassemble and status windows only update on trace mode
		on trace mode the disassemble window will follow the pc;
	- new disassemble implemented - function key to force another disassemble of
	the memory state.
* refactored IO thread to support new function keys. vm input is totally
separated from program input

dev version - 15/12/2015
------------------------------
* refactored curses.c for code readability
* threaded IO - a thread that delegates the io to the vm or the program being run
* curses interface - work in progress - looking pretty so far



vm/vmb.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#define FILE_CHUNK 65536
#define STACK_CHUNK 65536

#define REG(x) (mem[x]-32768)
#define ARG(x) (mem[x] <= 32767 ? mem[x] : reg[mem[x]-32768])
#define SET_REG(x, y) (reg[mem[x]-32768] = y)

#define A ARG(pc+1)
#define B ARG(pc+2)
#define C ARG(pc+3)

uint16_t mem[65536];
uint16_t stack[65536];
uint16_t reg[8];
uint16_t pc;
int sp = 0;

// functions definition
// stop execution and terminate the program
void halt(uint16_t a, uint16_t b, uint16_t c);
//set register <a> to the value of <b>
void set(uint16_t a, uint16_t b, uint16_t c);
//push <a> onto the stack
void push(uint16_t a, uint16_t b, uint16_t c);
//remove the top element from the stack and write it into <a>; empty stack = error
void pop(uint16_t a, uint16_t b, uint16_t c);
//set <a> to 1 if <b> is equal to <c>; set it to 0 otherwise
void eq(uint16_t a, uint16_t b, uint16_t c);
//set <a> to 1 if <b> is greater than <c>; set it to 0 otherwise
void gt(uint16_t a, uint16_t b, uint16_t c);
//jump to <a>
void jmp(uint16_t a, uint16_t b, uint16_t c);
//if <a> is nonzero, jump to <b>
void jt(uint16_t a, uint16_t b, uint16_t c);
//if <a> is zero, jump to <b>
void jf(uint16_t a, uint16_t b, uint16_t c);
//assign into <a> the sum of <b> and <c> (modulo 32768)
void add(uint16_t a, uint16_t b, uint16_t c);
//store into <a> the product of <b> and <c> (modulo 32768)
void mult(uint16_t a, uint16_t b, uint16_t c);
//store into <a> the remainder of <b> divided by <c>
void mod(uint16_t a, uint16_t b, uint16_t c);
//stores into <a> the bitwise and of <b> and <c>
void and(uint16_t a, uint16_t b, uint16_t c);
//stores into <a> the bitwise or of <b> and <c>
void or(uint16_t a, uint16_t b, uint16_t c);
//stores 15-bit bitwise inverse of <b> in <a>
void not(uint16_t a, uint16_t b, uint16_t c);
//read memory at address <b> and write it to <a>
void rmem(uint16_t a, uint16_t b, uint16_t c);
//write the value from <b> into memory at address <a>
void wmem(uint16_t a, uint16_t b, uint16_t c);
//write the address of the next instruction to the stack and jump to <a>
void call(uint16_t a, uint16_t b, uint16_t c);
//remove the top element from the stack and jump to it; empty stack = halt
void ret(uint16_t a, uint16_t b, uint16_t c);
//write the character represented by ascii code <a> to the terminal
void out(uint16_t a, uint16_t b, uint16_t c);
//read a character from the terminal and write its ascii code to <a>; it can be assumed that once input starts
//it will continue until a newline is encountered; this means that you can safely read whole the keyboard and 
//trust that they will be fully read
void in(uint16_t a, uint16_t b, uint16_t c);
// no operation
void noop(uint16_t a, uint16_t b, uint16_t c);

void (*opcode_function[22])() = {
  halt, set, push,
  (int *)&pop, (int *) &eq, (int *) &gt,
  (int *)&jmp, (int *) &jt, (int *) &jf,
  (int *)&add, (int *) &mult, (int *) &mod,(int*) &and,
  (int *)&or, (int *) &not, (int *) &rmem,
  (int *)&wmem, (int *) &call, (int *) &ret,
  (int *)&out, (int *) &in, (int *) &noop
};

void halt(uint16_t a, uint16_t b, uint16_t c) {
  exit(0);
}

void set(uint16_t a, uint16_t b, uint16_t c) {
  SET_REG(pc + 1, b);
  pc += 3;
}

void push(uint16_t a, uint16_t b, uint16_t c) {
  stack[sp++] = a;
  pc += 2;
}

void pop(uint16_t a, uint16_t b, uint16_t c) {
  if (!sp) halt(a, b, c);
  sp--;
  pc = stack[sp];
}

void eq(uint16_t a, uint16_t b, uint16_t c) {
  SET_REG(pc + 1, b == c);
  pc += 4;
}

void gt(uint16_t a, uint16_t b, uint16_t c) {
  SET_REG(pc + 1, b > c);
  pc += 4;
}

void jmp(uint16_t a, uint16_t b, uint16_t c) {
  pc = a;
}

void jt(uint16_t a, uint16_t b, uint16_t c) {
  pc = a ? b : pc + 3;
}

void jf(uint16_t a, uint16_t b, uint16_t c) {
  pc = !a ? b : pc + 3;
}

void add(uint16_t a, uint16_t b, uint16_t c) {
  SET_REG(pc + 1, (b + c) % 32768);
  pc += 4;
}

void mult(uint16_t a, uint16_t b, uint16_t c) {
  SET_REG(pc + 1, (b * c) % 32768);
  pc += 4;
}

void mod(uint16_t a, uint16_t b, uint16_t c) {
  SET_REG(pc + 1, (b % c));
  pc += 4;
}

void and(uint16_t a, uint16_t b, uint16_t c) {
  SET_REG(pc + 1, (b & c));
  pc += 4;
}

void or(uint16_t a, uint16_t b, uint16_t c) {
  SET_REG(pc + 1, (b | c));
  pc += 4;
}

void not(uint16_t a, uint16_t b, uint16_t c) {
  SET_REG(pc + 1, b^0x7FFF);
  pc += 3
}

void rmem(uint16_t a, uint16_t b, uint16_t c) {
  exit(0);
}

void wmem(uint16_t a, uint16_t b, uint16_t c) {
  exit(0);
}

void call(uint16_t a, uint16_t b, uint16_t c) {
  exit(0);
}

void ret(uint16_t a, uint16_t b, uint16_t c) {
  exit(0);
}

void out(uint16_t a, uint16_t b, uint16_t c) {
  putchar(a);
  pc += 2;
}

void in(uint16_t a, uint16_t b, uint16_t c) {
  exit(0);
}

void noop(uint16_t a, uint16_t b, uint16_t c) {
  pc++;
}



// load the program to the vm memory
void load(char * file) {
  char ch;
  int fd = open(file, O_RDONLY);

  if (fd == -1) {
    printf("Error loading program [%s]\n", file);
    exit(1);
  }

  ssize_t br, offset = 0;
  while((br = read(fd, mem + offset, sizeof(uint16_t))) > 0) {
    offset++;
  }

  printf("Loaded program with %d bytes\n", offset); 

  close(fd);
}

// runs the program. return 0 for successfull run
// and != 0 for error
int run() {
  while(1) {
    uint16_t opcode = mem[pc];

    if (!opcode) break;

    //    printf("%d %d \n", sp, opcode);
    if (opcode == 19) {
      putchar((char) (ARG(pc +1)));
      pc+=2;
    } else if (opcode == 21) {
      pc++;
    } else if (opcode == 6) {
      pc = ARG(pc+1);
    } else if (opcode == 7) {
      pc = ARG(pc+1) ? ARG(pc+2) : pc + 3;
    } else if (opcode == 8) {
      pc = !ARG(pc+1) ? ARG(pc+2) : pc + 3;
    } else if (opcode == 1) {
      SET_REG(pc + 1, ARG(pc + 2));
      pc += 3;
    } else if (opcode == 9) {
      SET_REG(pc + 1, (ARG(pc+2) + ARG(pc+3))%32768);
      pc += 4;
    } else if (opcode == 4) {
      SET_REG(pc + 1, ARG(pc + 2) == ARG(pc + 3));
      pc += 4;
    } else if (opcode == 2) {
      stack[sp++] = ARG(pc + 1);
      pc += 2;
    } else if (opcode == 3) {
      sp--;
      SET_REG(pc + 1, stack[sp]);
      pc += 2;
    } else if(opcode == 5) {
      SET_REG(pc + 1, ARG(pc + 2) > ARG(pc + 3));
      pc += 4;
    } else if (opcode == 10) {
      SET_REG(pc + 1, (ARG(pc+2) * ARG(pc+3))%32768);
      pc += 4;
    } else if (opcode == 11) {
      SET_REG(pc + 1, (ARG(pc+2) % ARG(pc+3)));
      pc += 4;
    } else if (opcode == 12) {
      SET_REG(pc + 1, (ARG(pc+2) & ARG(pc+3)));
      pc += 4;
    } else if (opcode == 13) {
      SET_REG(pc + 1, (ARG(pc+2) | ARG(pc+3)));
      pc += 4;
    } else if (opcode == 14) {
      SET_REG(pc + 1, ARG(pc +2)^0x7FFF);
      pc += 3;
    } else if (opcode == 17) {
      stack[sp++] = pc+2;
      pc = ARG(pc+1);
    } else if (opcode == 15) {
      SET_REG(pc + 1, mem[ARG(pc+2)]);
      pc +=3;
    } else if (opcode == 16) {
      mem[ARG(pc+1)] = ARG(pc+2);
      pc +=3;
    } else if (opcode == 18) {
      if (!sp) break;
      sp--;
      pc = stack[sp];

    } else if (opcode = 20) {
      SET_REG(pc + 1, getchar());
      pc+=2;
    } else {
      pc++;
    }
  }
}

int main(int argc, char **argv) {
  if (argc != 2) {
    printf("usage: %s <program_to_load>\n", argv[0]);
    exit(1);
  }

  (opcode_function[1])();
  //load the program
  load(argv[1]);

  //run the program
  run();
}
